### PR TITLE
cmake: allow universal build on macOS 10.13

### PR DESCRIPTION
**Pull request by MarcusCalhoun-Lopez**
###### Type(s)

- [ ] bugfix
- [ ] enhancement
- [ ] security fix

###### Tested on

macOS 10.13
Xcode 10.1

###### Verification 
Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [x] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL?

- [x] checked your Portfile with `port lint`?
- [x] tried existing tests with `sudo port test`?
- [x] tried a full install with `sudo port -vst install`?
- [x] tested basic functionality of all binary files?

### CODE REVIEW

**Comment by michaelld** [comment]
@kencu can you test this one? My 10.13 box is under the weather for a bit ...

**Comment by kencu** [comment]
Sure. For that last error, the linker warning, I used:
```
configure.ldflags-append -Wl,-w
```
instead of Marcus' approach. Simpler, or more complicated, I leave it up to Marcus to decide.

**Comment by MarcusCalhoun-Lopez** [comment]
I like the idea of using `configure.ldflags-append -Wl,-w`.
It certainly seems easier.
My only concern is that it would also suppress (perhaps useful) debug info for the entire build.
The patch file localized the suppression to the testing of features.

**Comment by kencu** [comment]
Valid point.

I'm trying to fully grasp what commit 2 is supposed to fix; building cmake +universal against the 10.13 SDK on Mojave, I had no issue with the way things are already. So I haven't really run across a problem so far.

I do think that base should always set SDKROOT to something, even "/" as a last resort, though.

I totally agree that we should completely bypass cmake's attempt to guess and find the proper SDK to use, as that is a doomed adventure on MacPorts, where "we know better".

Still reading...

**Comment by MarcusCalhoun-Lopez** [comment]
> I'm trying to fully grasp what commit 2 is supposed to fix; building cmake +universal against the 10.13 SDK on Mojave, I had no issue with the way things are already. So I haven't really run across a problem so far.

This is only a guess, but Xcode 10.1 is installed on my macOS 10.13 VM.
As a result, only the 10.14 SDK is installed.
Since `SDKROOT` was not set, the build system found the 10.14 SDK, which does not have 32-bit libraries, and failed.
If someone has Xcode 9.x installed, they might not see the same error.

**Comment by kencu** [comment]
So the issue is that if you're on MacOS 10.13, and you want to build universal, and you have Xcode 10.1 that has only the 10.14 SDK, which is not universal, and you don't want to install the MacOSX10.13.sdk to do it the official way, you would like MacPorts to use "/" as the SDK instead.

That is a good corner case -- I get it -- I'm not immediately certain that changing the way that every cmake build in MacPorts is done is the way to go though. To be honest, I am trying to think through all the variations regarding the myriad of ways this might go sour.

Also, this same issue will apply to every universal build with that setup, not just cmake, so wouldn't you have to do something like that in every single build?

We always want the SDK chosen by the user respected. And then whatever MacPorts chooses. And then probably "/" if we get to this point. We certainly don't want to leave it up to cmake to sort out.

I'm thinking we tell the user to install  the MacOSX10.13.sdk and use that, if they want to build universal on 10.13 with Xcode 10+. And probably do that in base? That's exactly what I did, on Mojave, building +universal against the 10.13 SDK.

**Comment by MarcusCalhoun-Lopez** [comment]
> So the issue is that if you're on MacOS 10.13, and you want to build universal, and you have Xcode 10.1 that has only the 10.14 SDK, which is not universal, and you don't want to install the MacOSX10.13.sdk to do it the official way, you would like MacPorts to use "/" as the SDK instead.
> 
> That is a good corner case -- I get it -- I'm not immediately certain that changing the way that cmake is built on every system in MacPorts is the way to go though. To be honest, I am trying to think through all the variations regarding the myriad of ways this might go sour.

This issues it not just limited to one system with one variant.
The [current code](https://github.com/macports/macports-ports/blob/8fc7886c4edfa792c56f58a053b0f223ebe5279f/devel/cmake/Portfile#L147) is not doing what it purports to to on **any** system.
It is just that on most systems, it is not so much of a problem.


> Also, this same issue will apply to every universal build with that setup, not just cmake, so wouldn't you have to do something like that in every single build?

We already do.
In the two cmake port groups, there is `configure.args-append -DCMAKE_OSX_SYSROOT="/"`.
Ports that use configure/make are fine since no SDK information is passed to the compiler (since `/usr/include` exists).

> I'm thinking we tell the user to install the MacOSX10.13.sdk and use that, if they want to build universal on 10.13 with Xcode 10+. And probably do that in base? That's exactly what I did, on Mojave, building +universal against the 10.13 SDK.

Even if we did this, we would still have to apply some sort of change like the one proposed here since, as you note, we do not want cmake [to find the wrong SDK](https://gitlab.kitware.com/cmake/cmake/blob/master/Modules/Platform/Darwin-Initialize.cmake#L65).

**Comment by kencu** [comment]
I don't know, Marcus. Maybe you're right. However, I just did this in the cmake portfile
```
configure.ldflags-append -Wl,-w
```
as I described here last year <https://trac.macports.org/ticket/56991>

and I copied the MacOSX10.13.sdk to where it belongs:
```
$ pwd
/Library/Developer/CommandLineTools/SDKs
Kens-MacBook-Pro:SDKs$ ls -la
total 0
drwxr-xr-x  5 root      wheel  160 23 Nov 17:24 .
drwxr-xr-x  5 root      admin  160  6 Oct  2017 ..
drwxr-xr-x  7 root      wheel  224  3 Nov  2018 MacOSX.sdk
drwxr-xr-x  5 root      wheel  160  9 Oct 23:54 MacOSX10.13.sdk
lrwxr-xr-x  1 root      wheel   10  3 Nov  2018 MacOSX10.14.sdk -> MacOSX.sdk
```

and everything works perfectly as it should building +universal without any other surgery. 
```
$ port -v installed cmake
The following ports are currently installed:
  cmake @3.15.5_0+universal (active) platform='darwin 17' archs='i386 x86_64' date='2019-11-23T17:49:33-0800'
```
So IMHO that is the right way to fix this issue, but at this point I defer to you, and @jmroot and @ryandesign to point out to me why I'm once again terribly wrong.

@michaelld -- back to you.

**Comment by michaelld** [comment]
Thanks @kencu . I'll need to get my 10.13 boot up and going & see what's going on here.

Does our CI check for +universal? A quick look at the Details doesn't provide much use info (to me) ...

**Comment by michaelld** [comment]
I greatly prefer not disabling linker warnings unless there's a very good reason to. So, instead of tweaking all linking such as what Ken proposes; and instead of the tweaking just the feature tests linking as what is proposed here; if the issue is on this exact setup: OSX 10.13, Xcode 10.1, and +universal ... then is it reasonable to narrow the patch to just this corner case & not patch otherwise?

**Comment by ryandesign** [comment]
> If someone has Xcode 9.x installed, they might not see the same error.

Right, I have no problem building universal on 10.13 with Xcode 9. That's the main reason I haven't upgraded my 10.13 machine or our 10.13 buildbot worker to Xcode 10.


> Does our CI check for +universal? A quick look at the Details doesn't provide much use info (to me) ...

Our CI builds and our buildbot builds only build default variants.


> I defer to you, and @jmroot and @ryandesign

I'm not going to have time to study this issue thoroughly enough to provide a meaningful opinion.

**Comment by MarcusCalhoun-Lopez** [comment]
> I greatly prefer not disabling linker warnings unless there's a very good reason to. So, instead of tweaking all linking such as what Ken proposes; and instead of the tweaking just the feature tests linking as what is proposed here; if the issue is on this exact setup: OSX 10.13, Xcode 10.1, and +universal ... then is it reasonable to narrow the patch to just this corner case & not patch otherwise?

As requested, I have tried to limit when the patch is applied.
I am not sure when the link warning started to appear (Xcode 10.0 or Xcode 10.1).

File: devel/cmake/Portfile [hunk 1]
```diff
@@ -104,6 +109,10 @@ if {${subport} eq ${name}} {
         patch-fix_cxx14_17_checks.devel.diff \
         patch-fix-system-prefix-path.devel.diff
 
+    if {${os.major}==17 && (${build_arch} eq "i386" || [variant_isset universal]) && [vercmp ${developerversion} 10] >= 0} {
+        patchfiles-append patch-fix-feature-test.devel.diff
```
Comment (kencu) [inline devel/cmake/Portfile]: the same error is also happening to folks running os.major=18 btw, like @Gcenx , so you might make him happy if you also handled that, as he's been asking me to fix this for months

Comment (kencu) [reply devel/cmake/Portfile]: of course, that's another corner case, and we can only go so far out of our way to make these corner case folks happy without making ourselves crazy in the process.

Comment (michaelld) [reply devel/cmake/Portfile]: @kencu do you think that are we better off with these "if" statements for corner cases, or does it hurt to just patch for every install? I can honestly go either way, especially here, since sometimes a massive patch is more useful even if it means we give up some info for when linking has warnings or the like.

Comment (Gcenx) [reply devel/cmake/Portfile]: @kencu if it’s changed to >=17 that would also cover Mojave universal and since base by default disabled universal above Mojave it wouldn’t cause any issues.

But this only covers cmake I’d also encountered issues with other cmake group ports. So I went back to patching base.

Comment (kencu) [reply devel/cmake/Portfile]: @michaelld -- I would say if you go with the patchfile to supress the i386 warning, it could be applied always, as it only cuts out that one warning, I believe. 

if you go with a system test, I would do this in the portfile (pseudocode):
```
if Xcode >= 10 and universal
   configure.ldflags-append -Wl,-w
end
```
as that will be resilient forever, whatever Cmake does in the future, whereas the patch will break eventually, someday, and I just hate patching unnecessarily. But that is me.

Regarding  the SDK, I would not change what the portfile does now, as it has survived years and years of trials. I would tell anyone who wants to build universal with Xcode 10+ that they have to install the MacOSX10.13.sdk and then build against it. On 10.13, building against the MacOSX10.13.sdk happens automagically as that is how base is written (as my test above). On 10.14, you have to do it manually, but either way, base already does the right thing and I don't believe we want to go backwards and make macports starting building against "/" at this point in time.

If anyone thinks that asking someone who wants to build universal on 10.13 with Xcode10+ to install the MacOSX10.13.sdk is too much to ask, I would not agree with that, as anyone doing that should darned well know just what they are doing.

Comment (kencu) [reply devel/cmake/Portfile]: One thing to ponder -- as @Gcenx points out -- it's possible this same linker warning breaks other universal builds later on that use cmake, in the same way. I haven't yet come across those, but if that were the case (I assume it is, as @Gcenx is not commonly incorrect) and if Marcus' patchfile were to also remove this warning from builds using cmake, not just the build of cmake itself, then his patch would be 100% the right way to go.

Also, if you don't mind watching for patch breakage over the next few years, Marcus' patch is definitely cleaner in that it removes only that one warning.

Comment (MarcusCalhoun-Lopez) [reply devel/cmake/Portfile]: > Regarding the SDK, I would not change what the portfile does now, as it has survived years and years of trials.

I am reasonably sure the SDK code needs to be fixed.
Perhaps I should not have included it in this pull request because it is **independent** of the universal variant.
The problem just came to light when I happen to be messing around with a universal build.

I am sorry if I was not clear before.
The code `configure.env-append CMAKE_OSX_SYSROOT=...` does nothing.
Perhaps it did at some point but  [no longer](https://gitlab.kitware.com/cmake/cmake/blob/master/Modules/Platform/Darwin-Initialize.cmake#L45).
If we want to set CMAKE_OSX_SYSROOT (which I would argue we should), we must set the environmental variable `SDKROOT`.
MacPorts mostly takes care of that for us, unless `${configure.sdkroot}` is empty, which is what the new code in the pull request handles.

Please let me know if I am still not being clear.

Comment (MarcusCalhoun-Lopez) [reply devel/cmake/Portfile]: > Marcus' patchfile were to also remove this warning from builds using cmake, not just the build of cmake itself ...

I believe that the patch affects only one small part of the build of cmake and not any ports that use cmake.

Comment (Gcenx) [reply devel/cmake/Portfile]: @MarcusCalhoun-Lopez CMAKE_OSX_SYSROOT doesn’t need to be set when building cmake only when using cmake to build something.

Comment (MarcusCalhoun-Lopez) [reply devel/cmake/Portfile]: If we do not set CMAKE_OSX_SYSROOT, then cmake searches for its own.
This is definitely not what we want.
Building any port should, as much as possible, respect the MacPorts configuration.

Comment (kencu) [reply devel/cmake/Portfile]: But the cmake portfile already sets this correctly, to ${configure.sdkroot}, which is perfect.

What am I missing?

Comment (kencu) [reply devel/cmake/Portfile]: I think the error Marcus saw in building cmake +universal was caused by base setting the SDK to MacOSX.sdk, which is the fallback it will go to if MacOSX10.13.sdk doesn't exist (on 10.13). The proper fix for that, IMHO, is to install the MacOSX10.13.sdk.

I feel like I must be looking at something sideways here.

Comment (Gcenx) [reply devel/cmake/Portfile]: Setting CMAKE_OSX_SYSROOT to configure.sdk also sets sdkroot so the change isn’t required, cmake “should” respect sdkroot but it might no in some cases? That’s lightly why the original Portfile was setting CMAKE_OSX_SYSROOT.

That would also ensure cmake doesn’t try overwriting sdkroot with the sdk it selected

Comment (MarcusCalhoun-Lopez) [reply devel/cmake/Portfile]: Please forgive my inability to properly convey the issue.
Also, please forgive me if, as a result, my explanation is overly verbose.
1. The current Portfile code (`configure.env-append CMAKE_OSX_SYSROOT=...`) does **nothing**&nbsp;[1]. At the very least we should remove it.
1. When building cmake, cmake attempts to find a value for `CMAKE_OSX_SYSROOT`. It guesses as best it can&nbsp;[1].
1. When building cmake, cmake adds `-isysroot ${CMAKE_OSX_SYSROOT}` to the compiler flags&nbsp;[2].
1. The value of `CMAKE_OSX_SYSROOT`, whatever it is, is **not** recorded in the installation. In other words, no port that uses cmake is affected.

> Setting CMAKE_OSX_SYSROOT to configure.sdk also sets sdkroot

> But the cmake portfile already sets this correctly

I do not believe this is correct.
I was unable to get this behavior, I could not any reference to `$ENV{CMAKE_OSX_SYSROOT}`, and the there is no indication of this behavior in the cmake code&nbsp;[1].

> cmake to not select the sdk would cause problems if called outside of macports

Projects outside of MacPorts will be unaffected by setting `SDKROOT` during the **build** of cmake.
Even if they were, the base sets `SDKROOT` anyway.

> The proper fix for that, IMHO, is to install the MacOSX10.13.sdk.

We can certainly modify Xcode and command line tools until we find a way to have cmake guess the right value for `CMAKE_OSX_SYSROOT`, but it makes much more sense to set it to the value we want in the first place.

Please let me know if I am still not explaining the situation correctly.

[1] https://gitlab.kitware.com/cmake/cmake/blob/master/Modules/Platform/Darwin-Initialize.cmake#L42
[2] https://cmake.org/cmake/help/v3.15/variable/CMAKE_OSX_SYSROOT.html

Comment (Gcenx) [reply devel/cmake/Portfile]: > 1. The current Portfile code (`configure.env-append CMAKE_OSX_SYSROOT=...`) does **nothing** [1]. At the very least we should remove it.


```
    if {${configure.sdkroot} != ""} {
        configure.env-append CMAKE_OSX_SYSROOT=${configure.sdkroot}
    } else {
        configure.env-append CMAKE_OSX_SYSROOT=/
    }
```

The code does nothing unless macports-base sets an SDK.

If sdkroot is set in my case I'm forcing the 10.13SDK on Mojave, then I will get;
`CMAKE_OSX_SYSROOT=${configure.sdkroot}`
As sdkroot set your getting
`configure.env-append CMAKE_OSX_SYSROOT=/`
The issue is macports-base isn't setting an SDK in your case.

> 2. When building cmake, cmake attempts to find a value for `CMAKE_OSX_SYSROOT`. It guesses as best it can [1].

If that is not set then yes cmake will figure it out on it's own.
```cmake
if(CMAKE_OSX_SYSROOT)
  # Use the existing value without further computation to choose a default.
  set(_CMAKE_OSX_SYSROOT_DEFAULT "${CMAKE_OSX_SYSROOT}")
````

If not it will attempt to use `SDKROOT` to set `CMAKE_OSX_SYSROOT`
```cmake
elseif(NOT "x$ENV{SDKROOT}" STREQUAL "x" AND
        (NOT "x$ENV{SDKROOT}" MATCHES "/" OR IS_DIRECTORY "$ENV{SDKROOT}"))
  # Use the value of SDKROOT from the environment.
  set(_CMAKE_OSX_SYSROOT_DEFAULT "$ENV{SDKROOT}")
```

> 3. When building cmake, cmake adds `-isysroot ${CMAKE_OSX_SYSROOT}` to the compiler flags [2].

Seems normal for cmake that's cmake's own way of passing the SDK onto clang/gcc but clang can also just use SDKROOK if -isysroot isn't set on recent clang .

> 4. The value of `CMAKE_OSX_SYSROOT`, whatever it is, is **not** recorded in the installation. In other words, no port that uses cmake is affected.

You were talking about patching that out that check before, that would be bad.

> 
> > Setting CMAKE_OSX_SYSROOT to configure.sdk also sets sdkroot
> 
> > But the cmake portfile already sets this correctly
> 
> I do not believe this is correct.
> I was unable to get this behavior, I could not any reference to `$ENV{CMAKE_OSX_SYSROOT}`, and the there is no indication of this behavior in the cmake code [1].

Read my comment above.


> 
> > cmake to not select the sdk would cause problems if called outside of macports
> 
> Projects outside of MacPorts will be unaffected by setting `SDKROOT` during the **build** of cmake.
> Even if they were, the base sets `SDKROOT` anyway.

You were talking about patching that out that check before.


> > The proper fix for that, IMHO, is to install the MacOSX10.13.sdk.
> 
> We can certainly modify Xcode and command line tools until we find a way to have cmake guess the right value for `CMAKE_OSX_SYSROOT`, but it makes much more sense to set it to the value we want in the first place.

Have you tried placing the 10.13SDK into the location @kencu mention above?
> 
> Please let me know if I am still not explaining the situation correctly.
> 
> [1] https://gitlab.kitware.com/cmake/cmake/blob/master/Modules/Platform/Darwin-Initialize.cmake#L42
> [2] https://cmake.org/cmake/help/v3.15/variable/CMAKE_OSX_SYSROOT.html

Comment (MarcusCalhoun-Lopez) [reply devel/cmake/Portfile]: > ```cmake
> if(CMAKE_OSX_SYSROOT)
>   # Use the existing value without further computation to choose a default.
>   set(_CMAKE_OSX_SYSROOT_DEFAULT "${CMAKE_OSX_SYSROOT}")
> ```
> 
> If not it will attempt to use `SDKROOT` to set `CMAKE_OSX_SYSROOT`

Yes, I agree.
In fact, that the point I am (poorly) trying to make.
`CMAKE_OSX_SYSROOT` is not defined in the Portfile, no matter the system configuration.
`ENV{CMAKE_OSX_SYSROOT}` is [defined](https://cmake.org/cmake/help/latest/variable/ENV.html), but `CMAKE_OSX_SYSROOT` is not.
The Portfile sets the environmental variable `CMAKE_OSX_SYSROOT`, which is **not** the same thing as passing `-DCMAKE_OSX_SYSROOT=...` to cmake.
That is why `configure.env-append CMAKE_OSX_SYSROOT=...` does not do anything.

Please correct me if I am wrong here.

> > 1. When building cmake, cmake adds `-isysroot ${CMAKE_OSX_SYSROOT}` to the compiler flags [2].
> 
> Seems normal for cmake that's cmake's own way of passing the SDK onto clang/gcc but clang can also just use SDKROOK if -isysroot isn't set on recent clang .

Yes, but cmake **is** setting `-isysroot`, sometimes to the wrong value.

> 
> > 1. The value of `CMAKE_OSX_SYSROOT`, whatever it is, is **not** recorded in the installation. In other words, no port that uses cmake is affected.
> 
> You were talking about patching that out that check before, that would be bad.

Tha patch is for a completely unrelated issue.
I should have opened two different pull requests.
Sorry for the confusion.
The patch I proposed has no affect on `SDKROOT`, `CMAKE_OSX_SYSROOT`, `-isysroot`, or any ports that use cmake.

> Have you tried placing the 10.13SDK into the location @kencu mention above?

Yes, I have.
However, instead of modifying system files, it makes much more sense to simply set `CMAKE_OSX_SYSROOT` to the value we want, as the Portfile currently **tries** to do.

Comment (kencu) [reply devel/cmake/Portfile]: > ENV{CMAKE_OSX_SYSROOT} is defined, but CMAKE_OSX_SYSROOT is not.

Now I see your point. I'm sorry I'm so slow. I assumed that the cmake build picked up CMAKE_OSX_SYSROOT from the environment (otherwise why set it in the Portfile?) but I did not go back and challenge the original premise that it did. If it truly does not, then you're right, that whole part of the portfile that sets CMAKE_OSX_SYSROOT in the environment is pointless.


> Have you tried placing the 10.13SDK into the location @kencu mention above?
> 
> Yes, I have.
> However, instead of modifying system files, it makes much more sense to simply set CMAKE_OSX_SYSROOT to the value we want, as the Portfile currently tries to do.

You might tweak cmake to build universal without the MacOSX10.13.sdk in the proper place, but nothing else will build universal unless likewise hacked. It's trivially easy to put the SDK there, and I can't help but believe it's just the right thing to do.

**Comment by kencu** [comment]
The following is right out of the cmake PortGroup. Should the cmake Portfile use this instead?
```
        if {${configure.sdkroot} != ""} {
            configure.args-append -DCMAKE_OSX_SYSROOT="${configure.sdkroot}"
        } else {
            configure.args-append -DCMAKE_OSX_SYSROOT="/"
        }
```

**Comment by jmroot** [comment]
Yes, that's probably the correct way of doing it, Ken. It's normal for no SDK to be used when the Command Line Tools provide /usr/include.

**Comment by MarcusCalhoun-Lopez** [comment]
> The following is right out of the cmake PortGroup. Should the cmake Portfile use this instead?
> 
> ```
>         if {${configure.sdkroot} != ""} {
>             configure.args-append -DCMAKE_OSX_SYSROOT="${configure.sdkroot}"
>         } else {
>             configure.args-append -DCMAKE_OSX_SYSROOT="/"
>         }
> ```

Unfortunately, this will not work either (it is one of the first things I tried).
`-DCMAKE_OSX_SYSROOT=` is something only `cmake` understands.
The `configure` script that cmake uses to bootstrap a `cmake` binary does not understand it. 
The configure phase will [fali](https://gitlab.kitware.com/cmake/cmake/blob/master/bootstrap#L881) with error "Unknown option: ...".

The only way I have found to achieve the same **behavior** is the way I proposed in the pull request.

**Comment by Gcenx** [comment]
@MarcusCalhoun-Lopez as a precaution it might be better doing the following;
```
    if {${configure.sdkroot} != ""} {
        configure.env-append SDKROOT=${configure.sdkroot}
    } else {
        configure.env-append SDKROOT=/
    }
```

**Comment by MarcusCalhoun-Lopez** [comment]
> @MarcusCalhoun-Lopez as a precaution it might be better doing the following;
> 
> ```
>     if {${configure.sdkroot} != ""} {
>         configure.env-append SDKROOT=${configure.sdkroot}
>     } else {
>         configure.env-append SDKROOT=/
>     }
> ```

Is there any particular scenario in which duplicating [code in the base](https://github.com/macports/macports-base/blob/2249c806dafa35c0bd2b2bce0bec29c94fa79856/src/port1.0/portutil.tcl#L429) might be necessary?

**Comment by Gcenx** [comment]
> > @MarcusCalhoun-Lopez as a precaution it might be better doing the following;
> > ```
> >     if {${configure.sdkroot} != ""} {
> >         configure.env-append SDKROOT=${configure.sdkroot}
> >     } else {
> >         configure.env-append SDKROOT=/
> >     }
> > ```
> 
> Is there any particular scenario in which duplicating [code in the base](https://github.com/macports/macports-base/blob/2249c806dafa35c0bd2b2bce0bec29c94fa79856/src/port1.0/portutil.tcl#L429) might be necessary?

It’s lightly not required more of a fail safe since that’s what was currently being done.
Since I’m always forcing an SDK (macOS Mojave using 10.13SDK) I’ll test your current version after I’m done with work today and see if it works as is or not.

**Comment by michaelld** [comment]
So ... where are we on this PR? I don't mind temporary fixes for ports as important as cmake ... but I think something as "meta" as this issue is (building for 32-bit on modern Xcode that doesn't provide the SDK for such builds) deserves meta treatment -- like, in "port" itself ... eventually.

**Comment by kencu** [comment]
The only way that the general case of fixing universal builds on 10.13 with xcode 10+ can be accomplished is by adding the 10.13 sdk as I described above.

The linker warning still needs to be silenced in certain ports where it triggers a build error. In yhis port, Marcus' patch silences it. Internally, when cmake wants to turn off linker warnings on other builds that fail due to similar warnings, it passes -w to the linker. 

In the general case of building universal on 10.13 with xcode 10, I just always pass that linker flag to every build, in base, as life is short. But it can be tweaked individually in each port where it causes an error, as there are not that many that fail. In the general case, the -w flag is more useful.

So, the sdk part of this PR offers a narrow fix for only cmake, that IMHO is not overly useful to someone who actually wants to build ports univesal on 10.13 with xcode 10. 

Base could be modified to pass "/" for the sdkroot in the case of universal/10.13/xcode10.

Or just ask people to install the sdk, which just seems like the proper general fix to me, and base already handles this properly right now.

**Comment by MarcusCalhoun-Lopez** [comment]
> The only way that the general case of fixing universal builds on 10.13 with xcode 10+ can be accomplished is by adding the 10.13 sdk as I described above.

I do not believe this is correct.
All you have to do is set `CMAKE_OSX_SYSROOT` to `/`, which is what the Portfile currently **tries** to do, but which the pull request actually does.
Anything that uses cmake already sets `CMAKE_OSX_SYSROOT ` to `/` when necessary through the [cmake PortGroup](https://github.com/macports/macports-ports/blob/ae28ac24a743098bc9014da0d40bd98965a7ba41/_resources/port1.0/group/cmake-1.1.tcl#L435).

> The linker warning still needs to be silenced in certain ports where it triggers a build error. In yhis port, Marcus' patch silences it. Internally, when cmake wants to turn off linker warnings on other builds that fail due to similar warnings, it passes -w to the linker.

When cmake is checking for C++ features, it doe not pass `-w` to the linker.
It [filters out the warnings](https://gitlab.kitware.com/cmake/cmake/blob/master/Source/Checks/cm_cxx_features.cmake#L21)

> So, the sdk part of this PR offers a narrow fix for only cmake, that IMHO is not overly useful to someone who actually wants to build ports univesal on 10.13 with xcode 10.

The sdk part is not a narrow fix.
**Everyone** who builds cmake is using the SDK that cmake is choosing and not (necessarily) what MacPorts is requesting.
Further, there is code in the Portfile that is doing nothing.
The "sdk part" remedies the situation.

**Comment by MarcusCalhoun-Lopez** [comment]
> So ... where are we on this PR? I don't mind temporary fixes for ports as important as cmake ... but I think something as "meta" as this issue is (building for 32-bit on modern Xcode that doesn't provide the SDK for such builds) deserves meta treatment -- like, in "port" itself ... eventually.

There are three parts to this pull request.
Again, let me apologize for including them in the same pull request.

1) Replace `PortGroup cxx11 1.1` with `compiler.cxx_standard 2011`.
This does not seem to be controversial.
I think we can safely proceed with this change.
1) Silence linker warnings when checking for C++ features.
This rightly raises concerns.
I have opened an [upstream pull request](https://gitlab.kitware.com/cmake/cmake/merge_requests/4089).
There have been no comments yet, so we can wait until they have had a chance to weigh in.
1) Respect `${configure.sdkroot}`.
I must admit that I am a little confused as to why this is controversial.
Code in the Portfile attempts to respect `${configure.sdkroot}` but in fact does nothing.
This change fixes that.

Since this pull request has conflicts anyway, I propose that I create three pull requests.
That is what I should have done on the first place.
The pull request for issue 1 could be merged quickly.
The pull request for issue 2 could be merged/rejected based on what upstream has to say.
I do not know about the pull request for issue 3.
It seems to me to be a straightforward fix for an issue that affects everyone.
Part of the problem seems to be the title of the pull request.
Issue 3 is not just limited to universal.

I could keep this pull request for issue 3 since that is what most of the comments have been about.
How does that sound?

**Comment by kencu** [comment]
what I'm saying is that hacking a fix in for cmake doesn't fix the universal build of, say, wine, or 1000+ other ports.

you need to add the sdk anyway for a general fix for all ports. and then you don't need this PR's sdk fix.

**Comment by kencu** [comment]
If you are building cmake for example on 10.14, setting macports to use the 10.13 sdk, and building it universal, cmake most definitely uses the sdk macports sets. We've been doing that for more than a year in the wine on mojave ticket, and it works. so cmake is obeying base now.

**Comment by jmroot** [comment]
Marcus is 100% correct. Cmake is unusual in insisting on specifying an SDK even when /usr/include exists. So it is ignoring base's choice of "no SDK" unless you explicitly tell it to use "/".

**Comment by kencu** [comment]
base should always set an sdk, even if it is "/" in the end. this causes headaches elsewhere too.

**Comment by kencu** [comment]
if the cmake devs are willing to add a -w flag for the case of cmake building universal on 10.23 with xcode 10, I'll fall off my chair.

**Comment by kencu** [comment]
10.13, not 10.23...

**Comment by kencu** [comment]
10.13, not 10.23...

**Comment by MarcusCalhoun-Lopez** [comment]
> what I'm saying is that hacking a fix in for cmake doesn't fix the universal build of, say, wine, or 1000+ other ports.

I am afraid I have been unclear.
However, I am not sure how else to express the issue.
* The change to `CMAKE_OSX_SYSROOT` is not a hack.
* The universal build **revealed** a problem.
However, the problem affects everyone.

> you need to add the sdk anyway for a general fix for all ports. and then you don't need this PR's sdk fix.

Which we already do in the the [cmake PortGroup](https://github.com/macports/macports-ports/blob/a85eb9d08ea9a2fcbda5e6803e166158753c2346/_resources/port1.0/group/cmake-1.1.tcl#L435).
However, this pull request only addresses the **building** of cmake.
It has no affect on any ports that **use** cmake.

**Comment by MarcusCalhoun-Lopez** [comment]
> If you are building cmake for example on 10.14, setting macports to use the 10.13 sdk, and building it universal, cmake most definitely uses the sdk macports sets. We've been doing that for more than a year in the wine on mojave ticket, and it works. so cmake is obeying base now.

It is entirely possible that the algorithm that cmake uses happens to find, in certain circumstances, the same value as the one MacPorts desires.
That is **not** always the case.

**Comment by MarcusCalhoun-Lopez** [comment]
> base should always set an sdk, even if it is "/" in the end. this causes headaches elsewhere too.

That is beyond the scope of this pull request.

**Comment by MarcusCalhoun-Lopez** [comment]
> if the cmake devs are willing to add a -w flag for the case of cmake building universal on 10.23 with xcode 10, I'll fall off my chair.

Neither this pull request nor the upstream pull request is suggesting adding `-w`.
The upstream pull request filters out warnings from the linker only during the testing of C++ features (e.g. support for C++11).
The reason for this is that the feature test considers warnings as evidence that the features is **not** supported.

**Comment by kencu** [comment]
cmake already uses the -w flag now to silence unwanted linker warnings on other systems, which is why I pointed that out.

if base properly set an sdk always, this, and dozens of similar hacks, would not be needed, 
i believe. That's why it's in scope.

I'll stay with the only proper fix for building universal on 10.13 with xcode10 being to add the sdk. I've been doing it a long time now, and I know it works for yhe general case of all ports (modulo the linker warning issue).

**Comment by kencu** [comment]
> Which we already do in the the [cmake PortGroup](https://github.com/macports/macports-ports/blob/a85eb9d08ea9a2fcbda5e6803e166158753c2346/_resources/port1.0/group/cmake-1.1.tcl#L435).

only for ports that use cmake! The other non-cmake ports, and there are thousands, need the sdk anyway.

**Comment by jmroot** [comment]
I really get the feeling that you're talking about some other situation entirely, Ken. When building on 10.13 for 10.13, the toolchain will always look in / first without being told.

**Comment by kencu** [comment]
Apparently :> As you and Marcus both feel I have, I must have something basic backwards. You guys wrote it, after all, so must be me somehow.

**Comment by michaelld** [comment]
To be fair to Brad / CMake devs: they have basically been on vacation since the 3.16.0 release (Nov 26). I'm guessing they'll be back at it today, and hopefully there will be some traction on your PR.

**Comment by michaelld** [comment]
I'm going ahead with the C++11 change & the patches for the test warnings. Will just do those locally & push them directly. I have no issues with those, as they positively impact current functionality and otherwise don't really change anything.

**Comment by michaelld** [comment]
I honestly don't know about the SDK stuff. I need to get my older build systems going again in order to do testing. Not sure when that's going to happen, given my current workload.

**Comment by kencu** [comment]
I will have to go back and reassess my understanding of how base selects an SDK and how that gets uptake by ports during builds. Although I have been following this -- I thought -- fairly closely, it seems I must still have some knowledge gaps and that is a bit unsettling.

I thought what happened was that base used a specified versioned SDK if one was specified by the portfile or macports.conf and if that versioned SDK existed, and if not that then base picked a versioned SDK that matched the build OS version if that existed, and if not that then base fell back to an unversioned SDK (MacOSX.sdk) if that existed, and if none of those, then "/" would be used.

But it appears that is not what actually happens after all, and I'll have to reassess my premises. My apologies if my misunderstandings have caused any wasted time.

**Comment by kencu** [comment]
I installed a new clean 10.13 VM, with the CommandLineTools, and without Xcode. The CommandLineTools installed only the 10.14 SDK, symlinked to MacOS.sdk. And indeed, base does build against "/" in this configuration instead of against the 10.14 sdk, and things do build universal (except for cmake, which as pointed out does it's own thing, and fails, as will some other ports, I believe). So indeed, my understanding of how this was working appears incomplete.

I have opened a ticket for base to set SDKROOT and configure.sdkroot to "/" rather than just leaving them empty at the end of portconfigure, as I believe there is no need for every port interested in the sdkroot to have to check for "" and set "/" individually.

**Comment by Gcenx** [comment]
@kencu as long as that only applies to 10.13 and below it shouldn’t be an issue as 10.14 and above don’t have the required /usr/includes folder without running the additional pkg included within Xcode command line tools.

10.14 and above should always set an SDK as that’s the now expected behavior going forward.

**Comment by MarcusCalhoun-Lopez** [comment]
The [upstream pull request](https://gitlab.kitware.com/cmake/cmake/merge_requests/4089) has been merged, so I modified this pull request to apply the patch unconditionally.

Is there anything else I can do to help move this pull request along?

**Comment by kencu** [comment]
LGTM, Marcus.
